### PR TITLE
Drop debug info instead of panicking if we exceed LLVM's capability to represent it

### LINKED ISSUE #131944: [Regression] LLVM asserts "conflicting locations for variable" since 1.82
### Code

I tried this code: cargo itself, building release + debug=2
(multiple versions, but at the time of this report I have cargo `cf53cc54bb593b5ec3dc2be4b1702f50c36d24d5`)

I expected to see this happen: successful compilation

Instead, this happened:

In [Fedora 39 builds](https://koji.fedoraproject.org/koji/taskinfo?taskID=124928273) (with its LLVM 17), `rustc` hits a `libstdc++` assertion via LLVM:

```
/usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/optional:479: _Tp &std::_Optional_base_impl<llvm::DIExpression::FragmentInfo, std::_Optional_base<llvm::DIExpression::FragmentInfo>>::_M_get() [_Tp = llvm::DIExpression::FragmentInfo, _Dp = std::_Optional_base<llvm::DIExpression::FragmentInfo>]: Assertion 'this->_M_is_engaged()' failed.
rustc exited with signal: 6 (SIGABRT) (core dumped)
```

While investigating, I found that `llc` also crashes with the extracted IR:
https://bugzilla.redhat.com/show_bug.cgi?id=2226564#c10

In local builds with rust's LLVM 19 and assertions enabled, and also with CI "alt" builds, LLVM asserts "conflicting locations for variable":

```
.../cargo$ CARGO_PROFILE_RELEASE_DEBUG=2 cargo +e92993dbb43f0a5d17fe56e2d82f90435d6521c8-alt build --release
[...]
rustc: /checkout/src/llvm-project/llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp:303: void llvm::Loc::MMI::addFrameIndexExpr(const DIExpression *, int): Assertion `(FrameIndexExprs.size() == 1 || llvm::all_of(FrameIndexExprs, [](const FrameIndexExpr &FIE) { return FIE.Expr && FIE.Expr->isFragment(); })) && "conflicting locations for variable"' failed.
error: could not compile `cargo` (lib)
```

### Version it worked on

It most recently worked on: Rust 1.81.0

### Version with regression

Rust 1.82.0, but at first it only crashed in my Fedora 39 build with LLVM 17. However, using the bundled rust-llvm with assertions enabled does trigger a debuginfo assertion.

Current nightly (e92993dbb43f0a5d17fe56e2d82f90435d6521c8) in alt mode reproduces this as well.

i.e. `rustup-toolchain-install-master --alt e92993dbb43f0a5d17fe56e2d82f90435d6521c8`

```
$ rustc +e92993dbb43f0a5d17fe56e2d82f90435d6521c8-alt -Vv
rustc 1.84.0-nightly (e92993dbb 2024-10-18)
binary: rustc
commit-hash: e92993dbb43f0a5d17fe56e2d82f90435d6521c8
commit-date: 2024-10-18
host: x86_64-unknown-linux-gnu
release: 1.84.0-nightly
LLVM version: 19.1.1
```

### Backtrace

<details><summary><code>(gdb) backtrace</code></summary>
<p>

```
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0) at pthread_kill.c:44
#1  0x00007f70eb682793 in __pthread_kill_internal (threadid=<optimized out>, signo=6) at pthread_kill.c:78
#2  0x00007f70eb629d0e in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#3  0x00007f70eb611942 in __GI_abort () at abort.c:79
#4  0x00007f70eb61185e in __assert_fail_base (fmt=0x7f70eb7c5cb0 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n",
    assertion=assertion@entry=0x7f70db9545d4 <.L.str.70> "(FrameIndexExprs.size() == 1 || llvm::all_of(FrameIndexExprs, [](const FrameIndexExpr &FIE) { return FIE.Expr && FIE.Expr->isFragment(); })) && \"conflicting locations for variable\"",
    file=file@entry=0x7f70db8eb1f8 <str.67.llvm> "/checkout/src/llvm-project/llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp", line=line@entry=303,
    function=function@entry=0x7f70db233d86 <.L__PRETTY_FUNCTION__._ZN4llvm3Loc3MMI17addFrameIndexExprEPKNS_12DIExpressionEi> "void llvm::Loc::MMI::addFrameIndexExpr(const DIExpression *, int)") at assert.c:94
#5  0x00007f70eb621e47 in __assert_fail (
    assertion=0x7f70db9545d4 <.L.str.70> "(FrameIndexExprs.size() == 1 || llvm::all_of(FrameIndexExprs, [](const FrameIndexExpr &FIE) { return FIE.Expr && FIE.Expr->isFragment(); })) && \"conflicting locations for variable\"",
    file=0x7f70db8eb1f8 <str.67.llvm> "/checkout/src/llvm-project/llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp", line=303,
    function=0x7f70db233d86 <.L__PRETTY_FUNCTION__._ZN4llvm3Loc3MMI17addFrameIndexExprEPKNS_12DIExpressionEi> "void llvm::Loc::MMI::addFrameIndexExpr(const DIExpression *, int)") at assert.c:103
#6  0x00007f70e3710e95 in llvm::Loc::MMI::addFrameIndexExpr(llvm::DIExpression const*, int) [clone .cold] () from /home/jistone/.rustup/toolchains/e92993dbb43f0a5d17fe56e2d82f90435d6521c8-alt/lib/libLLVM.so.19.1-rust-1.84.0-nightly
#7  0x00007f70e22f2086 in llvm::DwarfDebug::collectVariableInfoFromMFTable(llvm::DwarfCompileUnit&, llvm::DenseSet<std::pair<llvm::DINode const*, llvm::DILocation const*>, llvm::DenseMapInfo<std::pair<llvm::DINode const*, llvm::DILocation const*>, void> >&) () from /home/jistone/.rustup/toolchains/e92993dbb43f0a5d17fe56e2d82f90435d6521c8-alt/lib/libLLVM.so.19.1-rust-1.84.0-nightly
#8  0x00007f70e25914ab in llvm::DwarfDebug::collectEntityInfo(llvm::DwarfCompileUnit&, llvm::DISubprogram const*, llvm::DenseSet<std::pair<llvm::DINode const*, llvm::DILocation const*>, llvm::DenseMapInfo<std::pair<llvm::DINode const*, llvm::DILocation const*>, void> >&) () from /home/jistone/.rustup/toolchains/e92993dbb43f0a5d17fe56e2d82f90435d6521c8-alt/lib/libLLVM.so.19.1-rust-1.84.0-nightly
#9  0x00007f70e2592fe8 in llvm::DwarfDebug::endFunctionImpl(llvm::MachineFunction const*) () from /home/jistone/.rustup/toolchains/e92993dbb43f0a5d17fe56e2d82f90435d6521c8-alt/lib/libLLVM.so.19.1-rust-1.84.0-nightly
#10 0x00007f70e267f0a4 in llvm::DebugHandlerBase::endFunction(llvm::MachineFunction const*) () from /home/jistone/.rustup/toolchains/e92993dbb43f0a5d17fe56e2d82f90435d6521c8-alt/lib/libLLVM.so.19.1-rust-1.84.0-nightly
#11 0x00007f70e2bb6059 in llvm::AsmPrinter::emitFunctionBody() () from /home/jistone/.rustup/toolchains/e92993dbb43f0a5d17fe56e2d82f90435d6521c8-alt/lib/libLLVM.so.19.1-rust-1.84.0-nightly
#12 0x00007f70e223d646 in llvm::X86AsmPrinter::runOnMachineFunction(llvm::MachineFunction&) () from /home/jistone/.rustup/toolchains/e92993dbb43f0a5d17fe56e2d82f90435d6521c8-alt/lib/libLLVM.so.19.1-rust-1.84.0-nightly
#13 0x00007f70e2d7ac94 in llvm::FPPassManager::runOnFunction(llvm::Function&) [clone .warm] () from /home/jistone/.rustup/toolchains/e92993dbb43f0a5d17fe56e2d82f90435d6521c8-alt/lib/libLLVM.so.19.1-rust-1.84.0-nightly
#14 0x00007f70e22d6376 in llvm::FPPassManager::runOnModule(llvm::Module&) () from /home/jistone/.rustup/toolchains/e92993dbb43f0a5d17fe56e2d82f90435d6521c8-alt/lib/libLLVM.so.19.1-rust-1.84.0-nightly
#15 0x00007f70e299f210 in llvm::legacy::PassManagerImpl::run(llvm::Module&) () from /home/jistone/.rustup/toolchains/e92993dbb43f0a5d17fe56e2d82f90435d6521c8-alt/lib/libLLVM.so.19.1-rust-1.84.0-nightly
#16 0x00007f70e9f2aab2 in LLVMRustWriteOutputFile () from /home/jistone/.rustup/toolchains/e92993dbb43f0a5d17fe56e2d82f90435d6521c8-alt/lib/librustc_driver-a4eb0db55e024726.so
#17 0x00007f70e9f2a767 in rustc_codegen_llvm::back::write::write_output_file () from /home/jistone/.rustup/toolchains/e92993dbb43f0a5d17fe56e2d82f90435d6521c8-alt/lib/librustc_driver-a4eb0db55e024726.so
#18 0x00007f70e9ef1d5c in rustc_codegen_llvm::back::write::codegen () from /home/jistone/.rustup/toolchains/e92993dbb43f0a5d17fe56e2d82f90435d6521c8-alt/lib/librustc_driver-a4eb0db55e024726.so
#19 0x00007f70e9ef19e2 in rustc_codegen_ssa::back::write::finish_intra_module_work::<rustc_codegen_llvm::LlvmCodegenBackend> () from /home/jistone/.rustup/toolchains/e92993dbb43f0a5d17fe56e2d82f90435d6521c8-alt/lib/librustc_driver-a4eb0db55e024726.so
#20 0x00007f70ea003a36 in std::sys::backtrace::__rust_begin_short_backtrace::<<rustc_codegen_llvm::LlvmCodegenBackend as rustc_codegen_ssa::traits::backend::ExtraBackendMethods>::spawn_named_thread<rustc_codegen_ssa::back::write::spawn_work<rustc_codegen_llvm::LlvmCodegenBackend>::{closure#0}, ()>::{closure#0}, ()> () from /home/jistone/.rustup/toolchains/e92993dbb43f0a5d17fe56e2d82f90435d6521c8-alt/lib/librustc_driver-a4eb0db55e024726.so
#21 0x00007f70ea000b28 in <<std::thread::Builder>::spawn_unchecked_<<rustc_codegen_llvm::LlvmCodegenBackend as rustc_codegen_ssa::traits::backend::ExtraBackendMethods>::spawn_named_thread<rustc_codegen_ssa::back::write::spawn_work<rustc_codegen_llvm::LlvmCodegenBackend>::{closure#0}, ()>::{closure#0}, ()>::{closure#1} as core::ops::function::FnOnce<()>>::call_once::{shim:vtable#0} ()
   from /home/jistone/.rustup/toolchains/e92993dbb43f0a5d17fe56e2d82f90435d6521c8-alt/lib/librustc_driver-a4eb0db55e024726.so
#22 0x00007f70ea000eeb in std::sys::pal::unix::thread::Thread::new::thread_start () from /home/jistone/.rustup/toolchains/e92993dbb43f0a5d17fe56e2d82f90435d6521c8-alt/lib/librustc_driver-a4eb0db55e024726.so
#23 0x00007f70eb680797 in start_thread (arg=<optimized out>) at pthread_create.c:447
#24 0x00007f70eb70478c in __GI___clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:78
```

</p>
</details>

I bisected this to #128861 (cc @khuey)
* The assertion fails with `3139ff09e9d07f7700f8d15ed25a231e29c43627-alt` (3139ff09e9d07f7700f8d15ed25a231e29c43627)
* Compilation succeeds with `026e9ed3f0c5b3ee6233fd23b5e497cb94cf6434-alt` (its parent 026e9ed3f0c5b3ee6233fd23b5e497cb94cf6434)

@rustbot modify labels: +regression-from-stable-to-stable -regression-untriaged

### LINKED ISSUE #132900: ICE: `Failed to encode discriminator in DILocation`
### Code

Can not reproduce it with minimized code for now, and the whole repository is private.

Update: Thanks for @buaazp providing a minimal reproduce codebase https://github.com/buaazp/rust-bug.

### Meta

`rustc --version --verbose`:
```
rustc 1.84.0-nightly (143ce0920 2024-11-10)
binary: rustc
commit-hash: 143ce0920a2307b19831160a01f06f107610f1b2
commit-date: 2024-11-10
host: x86_64-unknown-linux-gnu
release: 1.84.0-nightly
LLVM version: 19.1.3
```

### Error output

```
thread 'rustc' panicked at compiler/rustc_codegen_llvm/src/debuginfo/create_scope_map.rs:168:22:
Failed to encode discriminator in DILocation

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: compiler flags: --crate-type lib -C opt-level=3 -C linker-plugin-lto -C codegen-units=1 -C debuginfo=2 -Z threads=8 -C target-cpu=native

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
error: could not compile `***` (lib)
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
stack backtrace:
   0:     0x7f9ccd2b0ada - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::h2f1c5db75869eef0
   1:     0x7f9ccda47126 - core::fmt::write::h43b2df425572a034
   2:     0x7f9ccea4c991 - std::io::Write::write_fmt::h77e35077f9fad7c4
   3:     0x7f9ccd2b0932 - std::sys::backtrace::BacktraceLock::print::h0c19526194593f74
   4:     0x7f9ccd2b2e36 - std::panicking::default_hook::{{closure}}::h865cf38b8deeb5bb
   5:     0x7f9ccd2b2c80 - std::panicking::default_hook::h69c14be83dcb49b3
   6:     0x7f9ccc329849 - std[3c0d0ff6ea8c1a1f]::panicking::update_hook::<alloc[f87dac715322685f]::boxed::Box<rustc_driver_impl[83ab483273712ee2]::install_ice_hook::{closure#0}>>::{closure#0}
   7:     0x7f9ccd2b3548 - std::panicking::rust_panic_with_hook::h583f6aaba9a390bd
   8:     0x7f9ccd2b331a - std::panicking::begin_panic_handler::{{closure}}::h02a363aabf2442f1
   9:     0x7f9ccd2b0f89 - std::sys::backtrace::__rust_end_short_backtrace::h84ecae8072aa78ac
  10:     0x7f9ccd2b2fdc - rust_begin_unwind
  11:     0x7f9cc9d15640 - core::panicking::panic_fmt::h514b8325f5f2609d
  12:     0x7f9cca2aa44b - core::option::expect_failed::h3eb9ea9a180694f7
  13:     0x7f9cca085139 - rustc_codegen_llvm[163141f4e00b7ecb]::debuginfo::create_scope_map::make_mir_scope
  14:     0x7f9cce998306 - rustc_codegen_ssa[b2545927768e4a09]::mir::codegen_mir::<rustc_codegen_llvm[163141f4e00b7ecb]::builder::Builder>
  15:     0x7f9cce995b85 - rustc_codegen_llvm[163141f4e00b7ecb]::base::compile_codegen_unit::module_codegen
  16:     0x7f9cce96c4d1 - <rustc_codegen_llvm[163141f4e00b7ecb]::LlvmCodegenBackend as rustc_codegen_ssa[b2545927768e4a09]::traits::backend::ExtraBackendMethods>::compile_codegen_unit
  17:     0x7f9ccc1619ca - rayon[c458d3a2004a12bf]::iter::plumbing::bridge_producer_consumer::helper::<rayon[c458d3a2004a12bf]::vec::DrainProducer<(usize, &rustc_codegen_ssa[b2545927768e4a09]::assert_module_sources::CguReuse)>, rayon[c458d3a2004a12bf]::iter::filter_map::FilterMapConsumer<rayon[c458d3a2004a12bf]::iter::extend::ListVecConsumer, rustc_data_structures[dd92c1caaaf932d5]::sync::parallel::enabled::par_map<(usize, &rustc_codegen_ssa[b2545927768e4a09]::assert_module_sources::CguReuse), alloc[f87dac715322685f]::vec::Vec<(usize, &rustc_codegen_ssa[b2545927768e4a09]::assert_module_sources::CguReuse)>, (usize, (rustc_codegen_ssa[b2545927768e4a09]::ModuleCodegen<rustc_codegen_llvm[163141f4e00b7ecb]::ModuleLlvm>, u64)), std[3c0d0ff6ea8c1a1f]::collections::hash::map::HashMap<usize, (rustc_codegen_ssa[b2545927768e4a09]::ModuleCodegen<rustc_codegen_llvm[163141f4e00b7ecb]::ModuleLlvm>, u64), rustc_hash[8e68038b4b9c783c]::FxBuildHasher>, rustc_codegen_ssa[b2545927768e4a09]::base::codegen_crate<rustc_codegen_llvm[163141f4e00b7ecb]::LlvmCodegenBackend>::{closure#6}::{closure#1}>::{closure#0}::{closure#0}>>
  18:     0x7f9cce96a340 - <rustc_codegen_llvm[163141f4e00b7ecb]::LlvmCodegenBackend as rustc_codegen_ssa[b2545927768e4a09]::traits::backend::CodegenBackend>::codegen_crate
  19:     0x7f9ccec310e7 - <rustc_interface[fd324feaaa946fb2]::queries::Linker>::codegen_and_build_linker
  20:     0x7f9cce907f6a - rustc_interface[fd324feaaa946fb2]::interface::run_compiler::<core[3a5812f1f6fa4837]::result::Result<(), rustc_span[8e1c5aac56eafb7b]::ErrorGuaranteed>, rustc_driver_impl[83ab483273712ee2]::run_compiler::{closure#0}>::{closure#1}
  21:     0x7f9ccc33497d - <rayon_core[b1c0e86ef95b0e5b]::job::StackJob<rayon_core[b1c0e86ef95b0e5b]::latch::LatchRef<rayon_core[b1c0e86ef95b0e5b]::latch::LockLatch>, <rayon_core[b1c0e86ef95b0e5b]::registry::Registry>::in_worker_cold<<rayon_core[b1c0e86ef95b0e5b]::thread_pool::ThreadPool>::install<rustc_interface[fd324feaaa946fb2]::util::run_in_thread_pool_with_globals<rustc_interface[fd324feaaa946fb2]::interface::run_compiler<core[3a5812f1f6fa4837]::result::Result<(), rustc_span[8e1c5aac56eafb7b]::ErrorGuaranteed>, rustc_driver_impl[83ab483273712ee2]::run_compiler::{closure#0}>::{closure#1}, core[3a5812f1f6fa4837]::result::Result<(), rustc_span[8e1c5aac56eafb7b]::ErrorGuaranteed>>::{closure#3}::{closure#0}::{closure#1}::{closure#0}, core[3a5812f1f6fa4837]::result::Result<(), rustc_span[8e1c5aac56eafb7b]::ErrorGuaranteed>>::{closure#0}, core[3a5812f1f6fa4837]::result::Result<(), rustc_span[8e1c5aac56eafb7b]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[3a5812f1f6fa4837]::result::Result<(), rustc_span[8e1c5aac56eafb7b]::ErrorGuaranteed>> as rayon_core[b1c0e86ef95b0e5b]::job::Job>::execute
  22:     0x7f9ccbe081ae - <rayon_core[b1c0e86ef95b0e5b]::registry::WorkerThread>::wait_until_cold
  23:     0x7f9ccbe05c17 - <rayon_core[b1c0e86ef95b0e5b]::registry::ThreadBuilder>::run
  24:     0x7f9ccc32c4e6 - <<crossbeam_utils[fa9bdfd2dd3a184c]::thread::ScopedThreadBuilder>::spawn<<rayon_core[b1c0e86ef95b0e5b]::ThreadPoolBuilder>::build_scoped<rustc_interface[fd324feaaa946fb2]::util::run_in_thread_pool_with_globals<rustc_interface[fd324feaaa946fb2]::interface::run_compiler<core[3a5812f1f6fa4837]::result::Result<(), rustc_span[8e1c5aac56eafb7b]::ErrorGuaranteed>, rustc_driver_impl[83ab483273712ee2]::run_compiler::{closure#0}>::{closure#1}, core[3a5812f1f6fa4837]::result::Result<(), rustc_span[8e1c5aac56eafb7b]::ErrorGuaranteed>>::{closure#3}::{closure#0}::{closure#0}, rustc_interface[fd324feaaa946fb2]::util::run_in_thread_pool_with_globals<rustc_interface[fd324feaaa946fb2]::interface::run_compiler<core[3a5812f1f6fa4837]::result::Result<(), rustc_span[8e1c5aac56eafb7b]::ErrorGuaranteed>, rustc_driver_impl[83ab483273712ee2]::run_compiler::{closure#0}>::{closure#1}, core[3a5812f1f6fa4837]::result::Result<(), rustc_span[8e1c5aac56eafb7b]::ErrorGuaranteed>>::{closure#3}::{closure#0}::{closure#1}, core[3a5812f1f6fa4837]::result::Result<(), rustc_span[8e1c5aac56eafb7b]::ErrorGuaranteed>>::{closure#0}::{closure#0}::{closure#0}, ()>::{closure#0} as core[3a5812f1f6fa4837]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  25:     0x7f9ccc32099e - std[3c0d0ff6ea8c1a1f]::sys::backtrace::__rust_begin_short_backtrace::<alloc[f87dac715322685f]::boxed::Box<dyn core[3a5812f1f6fa4837]::ops::function::FnOnce<(), Output = ()> + core[3a5812f1f6fa4837]::marker::Send>, ()>
  26:     0x7f9ccc32c012 - <<std[3c0d0ff6ea8c1a1f]::thread::Builder>::spawn_unchecked_<alloc[f87dac715322685f]::boxed::Box<dyn core[3a5812f1f6fa4837]::ops::function::FnOnce<(), Output = ()> + core[3a5812f1f6fa4837]::marker::Send>, ()>::{closure#1} as core[3a5812f1f6fa4837]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  27:     0x7f9cce973839 - std::sys::pal::unix::thread::Thread::new::thread_start::h664ba273de986494
  28:     0x7f9cc8c10fa3 - start_thread
                               at /build/reproducible-path/glibc-2.28/nptl/pthread_create.c:486:8
  29:     0x7f9cc8b4206f - __GI___clone
                               at /build/reproducible-path/glibc-2.28/misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:95
  30:                0x0 - <unknown>
```

</p>
</details>

[rustc-ice-2024-11-11T12_30_37-1345317.txt](https://github.com/user-attachments/files/17702650/rustc-ice-2024-11-11T12_30_37-1345317.txt)

### PR DESCRIPTION
**Pull request by khuey**
Recapping a bit of history here:

In #128861 I made debug info correctly represent parameters to inline functions by removing a fake lexical block that had been inserted to suppress LLVM assertions and by deduplicating those parameters.

LLVM, however, expects to see a single parameter _with distinct locations_, particularly distinct inlinedAt values on the DILocations. This generally worked because no matter how deep the chain of inlines it takes two different call sites in the original function to result in the same function being present multiple times, and a function call requires a non-zero number of characters, but macros threw a wrench in that in #131944. At the time I thought the issue there was limited to proc-macros, where an arbitrary amount of code can be generated at a single point in the source text.

In #132613 I added discriminators to DILocations that would otherwise be the same to repair #131944[^1]. This works, but LLVM's capacity for discriminators is not infinite (LLVM actually only allocates 12 bits for this internally). At the time I thought it would be very rare for anyone to hit the limit, but #132900 proved me wrong. In the relatively-minimized test case it also became clear to me that the issue affects regular macros too, because the call to the inlined function will (without collapse_debuginfo on the macro) be attributed to the (repeated, if the macro is used more than once) textual callsite in the macro definition.

This PR fixes the panic by dropping debug info when we exceed LLVM's maximum discriminator value. There's also a preceding commit for a related but distinct issue: macros that use collapse_debuginfo should in fact have their inlinedAts collapsed to the macro callsite and thus not need discriminators at all (and not panic/warn accordingly when the discriminator limit is exhausted).

Fixes #132900

r? @jieyouxu 

[^1]: Editor's note: `fix` is a magic keyword in PR description that apparently will close the linked issue (it's closed already in this case, but still).

### CODE REVIEW

File: compiler/rustc_codegen_llvm/src/debuginfo/create_scope_map.rs [hunk 1]
```diff
@@ -161,24 +175,32 @@ fn make_mir_scope<'ll, 'tcx>(
...
+                debug_scope.as_mut().unwrap().inlined_at = Some(inlined_at);
+            }
+            None => {
+                // LLVM has a maximum discriminator that it can encode (currently
+                // it uses 12 bits for 4096 possible values. If we exceed that
+                // there is little we can do but drop the debug info and warn.
+                debug_scope = None;
+                tracing::warn!(
+                    "Debug information was lost because LLVM discriminator limits were exceeded"
+                );
```
Comment (jieyouxu) [inline compiler/rustc_codegen_llvm/src/debuginfo/create_scope_map.rs:199]: Question: this is a `tracing::warn` that is for rustc developers. But shouldn't this be a user-visible warning?

File: tests/ui/debuginfo/debuginfo-inline-callsite-location-macro-1.rs [hunk 2]
```diff
@@ -0,0 +1,4124 @@
...
+        Something::default()
+    }
+}
+
+fn collapsed_debuginfo() {
+    // LLVM allows 12 bits for encoding discriminators, so 4096+1 calls to a
+    // MIR-inlined function will exhaust it. But because we're using
+    // collapse_debuginfo(yes) for this macro, we shouldn't need discriminators
+    // at all.
+    collapsed_macro!(1);
```
Comment (jieyouxu) [inline tests/ui/debuginfo/debuginfo-inline-callsite-location-macro-1.rs:23]: Question: can these macro calls be generated by another macro?

**Review by jieyouxu — COMMENTED** [review]
Thanks. This is certainly... interesting. I have two questions.

File: compiler/rustc_codegen_llvm/src/debuginfo/create_scope_map.rs (hunk 1 above)
Comment (khuey) [reply compiler/rustc_codegen_llvm/src/debuginfo/create_scope_map.rs:199]: Maybe. It's not particularly actionable to users and we don't warn in all the other cases where we don't emit debug info.

File: tests/ui/debuginfo/debuginfo-inline-callsite-location-macro-1.rs (hunk 2 above)
Comment (khuey) [reply tests/ui/debuginfo/debuginfo-inline-callsite-location-macro-1.rs:23]: As far as I know not without a proc-macro (e.g. what the seq_macro crate does).

File: compiler/rustc_codegen_llvm/src/debuginfo/create_scope_map.rs (hunk 1 above)
Comment (jieyouxu) [reply compiler/rustc_codegen_llvm/src/debuginfo/create_scope_map.rs:199]: > we don't warn in all the other cases where we don't emit debuginfo.

I feel like as a user I would want to have such a warning, but maybe that can get very noisy. Anyway, if we do want to warn when not emitting debuginfo then we should do so consistently, so not for this PR.

**Review by jieyouxu — APPROVED** [review]
Otherwise this LGTM, thanks. r=me once PR CI is green.

**Comment by jieyouxu** [comment]
I think tests are not build with debug logging (for good reason), so you cannot include that in the stderr snapshot

**Comment by khuey** [comment]
Alright, I dropped the warning and adjusted the test accordingly.

@bors r=@jieyouxu

**Comment by cuviper** [comment]
Nominating a backport to follow #132613 -- it seems we should consider these as a pair, backporting both or neither.

@rustbot label beta-nominated

**Comment by apiraino** [comment]
Beta backport declined as per compiler team [on Zulip](https://rust-lang.zulipchat.com/#narrow/channel/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202024-11-21/near/483722955), since we also declined #132613

@rustbot label -beta-nominated

**Pull request merged**